test/jdk/java/net/httpclient/lib/jdk/httpclient/test/lib/common/HttpServerAdapters.java

```diff
@@ -421,6 +425,50 @@ private void doHandle(HttpTestExchange t) throws IOException {
         }
     }
 
+    /**
+     * An {@link HttpTestHandler} that handles only HEAD and GET
+     * requests. If another method is used 405 is returned with
+     * an empty body.
+     * The response is always returned with fixed length.
+     */
+    public static class HttpHeadHandler implements HttpTestHandler {
+        final String responseBody;
+        public HttpHeadHandler() {
+            this("pâté de tête persillé");
+        }
+        public HttpHeadHandler(String responseBody) {
+            this.responseBody = Objects.requireNonNull(responseBody);
+        }
+
+        @Override
+        public void handle(HttpTestExchange t) throws IOException {
+            t.getRequestBody().readAllBytes();
+            String method = t.getRequestMethod();
+            switch (method) {
+                case "HEAD" -> {
+                    byte[] resp = responseBody.getBytes(StandardCharsets.UTF_8);
+                    if (t.getExchangeVersion() != HTTP_1_1) {
+                        // with HTTP/2 or HTTP/3 the server will not send content-length
+                        t.getResponseHeaders().addHeader("Content-Length", String.valueOf(resp.length));
+                    }
+                    t.sendResponseHeaders(200, resp.length);
+                }
+                case "GET" -> {
+                    byte[] resp = responseBody.getBytes(StandardCharsets.UTF_8);
+                    t.sendResponseHeaders(200, resp.length);
+                    try (var os = t.getResponseBody()) {
+                        os.write(resp);
+                    }
+                }
+                default -> {
+                    t.sendResponseHeaders(405, 0);
+                }
+            }
+            t.getResponseBody().close();
```

Comment:
It's a `HttpTestExchange`, so we don't have clear defined semantics for what happens when `HttpTestExchange.getResponseBody()` is invoked after the response body is already `close()`d previously. I think it might be better to move this `t.getResponseBody().close()` into the individual `case` blocks to avoid calling `t.getResponseBody()` after the `case GET` already closes the response body in its try-with-resources.

Reply:
> It would be a bug if getResponseBody().close() were not.

Sorry, my comment wasn't clear. What I meant was, I think a second call to `getResponseBody()` (in theory) could throw an exception if `close()` has already been called on the `OutputStream` returned by the previous call to `getResponseBody()`.

Reply:
Oh - thanks yes. Should be good now.